pilot/cmd/pilot-agent/config/config.go

```diff
@@ -59,7 +61,16 @@ func ConstructProxyConfig(meshConfigFile, serviceCluster, proxyConfigEnv string,
 		proxyConfig = *meshConfig.DefaultConfig
 	}
 
-	if concurrency != 0 {
+	if concurrency == 0 && role.Type == model.Router {
+		// If --concurrency is unset and not override in file, env, or annotations, we will automatically set this
+		// based on CPU limits.
```

Comment:
Why not for sidecars ?

Reply:
Sidecars and gateways are treated differently:
https://github.com/istio/istio/pull/32928/files